src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java

```diff
@@ -164,103 +156,44 @@ public boolean partitionExists(final String topic, final String partition) throw
-    }
-
     @Override
-    public void syncPostBatch(final String topicId, final List<BatchItem> batch) throws EventPublishingException {
-        batch.forEach(item -> Preconditions.checkNotNull(
+    public void syncPostBatch(final String topicId, final List<BatchItem> batches) throws RuntimeException {
+        batches.forEach(item -> Preconditions.checkNotNull(
                 item.getPartition(), "BatchItem partition can't be null at the moment of publishing!"));
         final Producer<String, String> producer = kafkaFactory.takeProducer();
-        final Map<BatchItem, CompletableFuture<Exception>> sendFutures = new HashMap<>();
+        final List<PartitionInfo> partitionInfos;
         try {
-            for (final BatchItem item : batch) {
-                item.setStep(EventPublishingStep.PUBLISHING);
-                sendFutures.put(item, publishItem(producer, topicId, item));
-            }
-            final CompletableFuture<Void> multiFuture = CompletableFuture.allOf(
-                    sendFutures.values().toArray(new CompletableFuture<?>[sendFutures.size()]));
-            multiFuture.get(createSendTimeout(), TimeUnit.MILLISECONDS);
-
-            // Now lets check for errors
-            final Optional<Exception> needReset = sendFutures.entrySet().stream()
-                    .filter(entry -> isExceptionShouldLeadToReset(entry.getValue().getNow(null)))
-                    .map(entry -> entry.getValue().getNow(null))
-                    .findAny();
-            if (needReset.isPresent()) {
-                LOG.info("Terminating producer while publishing to topic {} because of unrecoverable exception",
-                        topicId, needReset.get());
-                kafkaFactory.terminateProducer(producer);
-            }
-        } catch (final TimeoutException ex) {
-            failUnpublished(batch, "timed out");
-            throw new EventPublishingException("Error publishing message to kafka", ex);
-        } catch (final ExecutionException ex) {
-            failUnpublished(batch, "internal error");
-            throw new EventPublishingException("Error publishing message to kafka", ex);
-        } catch (final InterruptedException ex) {
-            Thread.currentThread().interrupt();
-            failUnpublished(batch, "interrupted");
-            throw new EventPublishingException("Error publishing message to kafka", ex);
+            partitionInfos = producer.partitionsFor(topicId);
         } finally {
             kafkaFactory.releaseProducer(producer);
         }
-        final boolean atLeastOneFailed = batch.stream()
-                .anyMatch(item -> item.getResponse().getPublishingStatus() == EventPublishingStatus.FAILED);
-        if (atLeastOneFailed) {
-            failUnpublished(batch, "internal error");
-            throw new EventPublishingException("Error publishing message to kafka");
-        }
+
+        batches.stream().forEach(batch -> batch.setBrokerId(partitionInfos.stream()
```

Comment:
Instead of going through the list of `partitionInfos` for every `batchItem`, I would suggest to build a Map of partition -> brokerId only once. This code is executed a lot and I think it would be more performant to access a Map instead of loop through for every batch item.